### PR TITLE
doc: Add "debugging" section to the nRF5340 User Guide.

### PR DESCRIPTION
**Pull request by grochu**
It seems useful to add a debugging section to describe how to debug the network code firmware.

### CODE REVIEW

File: doc/nrf/ug_nrf5340.rst [hunk 1]
```diff
@@ -582,6 +582,34 @@ See the following instructions.
...
+.. note::
+  When debugging if you want to reset the network core remember to reset and run code execution on the application core first.
+
+.. note::
+  You can also use alternative tools for debugging:
+ 
+  * SEGGER Ozone.
+  * GDB command line tool. 
+
+.. note::
```
Comment (grochu) [inline doc/nrf/ug_nrf5340.rst:611]: This is to be confirmed.

File: doc/nrf/ug_nrf5340.rst [hunk 2]
```diff
@@ -582,6 +582,44 @@ See the following instructions.
...
+
+     SEGGER Ozone debugging session for the nRF5340 network core
+
+.. note::
+  When debugging if you want to reset the network core remember to reset and run code execution on the application core first.
+
+.. note::
+  You can also use alternative tools for debugging:
+ 
+  * SEGGER Ozone.
```
Comment (KAGA164) [inline doc/nrf/ug_nrf5340.rst:618]: Should it be `Visual Studio Code` since you described debugging with SEGGER Ozone above ?

Comment (grochu) [reply doc/nrf/ug_nrf5340.rst]: Good point. I think this is already resolved through the doc review from @divipillai. The thing is, we don't want to describe the very details of setting up the 2 debug sessions with the recommended toolchains as this is either hard (VS Code, and they plan to automate it) or we are not sure it's possible (SES). So we mention those tools as an alternative.

File: doc/nrf/links.txt [hunk 3]
```diff
@@ -122,6 +122,7 @@
 .. _`Creating an application`: https://nrfconnect.github.io/vscode-nrf-connect/connect/create_app.html
 .. _`Building an application`: https://nrfconnect.github.io/vscode-nrf-connect/connect/build_app.html
 .. _`Testing an application`: https://nrfconnect.github.io/vscode-nrf-connect/connect/test_app.html
+.. _`Debugging an application`: https://nrfconnect.github.io/vscode-nrf-connect/debug_app.html
```
Comment (peknis) [inline doc/nrf/links.txt:125]:
```suggestion
.. _`Debugging an application`: https://nrfconnect.github.io/vscode-nrf-connect/connect/debug_app.html
```

File: doc/nrf/ug_nrf5340.rst [hunk 4]
```diff
@@ -625,6 +623,37 @@ Container for firmware update binaries
 The build system will automatically place both the application core and the network core update binaries (:file:`app_update.bin` and :file:`net_core_app_update.bin`) into a container package named :file:`dfu_application.zip`.
 This container package can be used by update tools to pass both images during the simultaneous update of multiple images.
 
+Debugging
+*********
+
+To debug the application core firmware, you need a single debug session.
+The debug session is set up as described in `Debugging nRF5340 with SES`_ or `Debugging an application`_ with Visual Studio Code.
```
Comment (peknis) [inline doc/nrf/ug_nrf5340.rst:630]:
```suggestion
Set up the debug session as described in `Debugging nRF5340 with SES`_ or `Debugging an application`_ with Visual Studio Code.
```

File: doc/nrf/ug_nrf5340.rst [hunk 5]
```diff
@@ -625,6 +623,37 @@ Container for firmware update binaries
...
+
+However, to debug the firmware running on the network core, you need to set up separate debug sessions for each core:
+
+* Application core.
+* Network core.
```
Comment (peknis) [inline doc/nrf/ug_nrf5340.rst:635]:
```suggestion
To debug the firmware running on the network core, you also need to set up separate debug sessions for the application core.
```

File: doc/nrf/ug_nrf5340.rst [hunk 6]
```diff
@@ -625,6 +623,37 @@ Container for firmware update binaries
...
+
+Perform the following steps to start a debugging session in network core:
+
+1. Set up each session for application and network core as mentioned in `Debugging nRF5340 with SES`_ or `Debugging an application`_ with Visual Studio Code.
```
Comment (peknis) [inline doc/nrf/ug_nrf5340.rst:639]:
```suggestion
1. Set up a session for the application core and network core as mentioned in `Debugging nRF5340 with SES`_ or `Debugging an application`_ with Visual Studio Code.
```

File: doc/nrf/ug_nrf5340.rst [hunk 7]
```diff
@@ -625,6 +623,37 @@ Container for firmware update binaries
...
+#. Select the appropriate CPU for debugging, either the nRF5340 application core or the nRF5340 network core.
+#. Once both sessions are established, run the code execution on the application core.
+   The startup code will release the ``NETWORK.FORCEOFF`` signal to start the network core and it will allocate the necessary GPIO pins for the network core.
```
Comment (peknis) [inline doc/nrf/ug_nrf5340.rst:642]:
```suggestion
#. Once both sessions are established, execute the code on the application core.

   The startup code releases the ``NETWORK.FORCEOFF`` signal to start the network core and allocates the necessary GPIO pins for it.
```

File: doc/nrf/ug_nrf5340.rst [hunk 8]
```diff
@@ -625,6 +623,37 @@ Container for firmware update binaries
...
+#. Start code execution on the network core in the other debug session.
+
```
Comment (peknis) [inline doc/nrf/ug_nrf5340.rst:644]: Three consecutive notes is not ok. I believe one note only should be enough. Maybe the last one.

File: doc/nrf/ug_nrf5340.rst [hunk 9]
```diff
@@ -625,6 +623,37 @@ Container for firmware update binaries
...
+.. note::
+  While debugging, if you want to reset the network core, ensure to first reset and run code execution on the application core.
```
Comment (peknis) [inline doc/nrf/ug_nrf5340.rst:646]:
```suggestion
  If you want to reset the network core while debugging, make sure to first reset the application core and execute the code.
```

File: doc/nrf/ug_nrf5340.rst [hunk 10]
```diff
@@ -625,6 +623,37 @@ Container for firmware update binaries
...
+
+.. note::
+  You can use alternative tools for debugging:
```
Comment (peknis) [inline doc/nrf/ug_nrf5340.rst:649]:
```suggestion
  You can use, for example, the following tools for debugging:
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: doc/nrf/ug_nrf5340.rst [hunk 11]
```diff
@@ -625,6 +623,37 @@ Container for firmware update binaries
...
+
+To debug the application core firmware, you need a single debug session.
+The debug session is set up as described in `Debugging nRF5340 with SES`_ or `Debugging an application`_ with Visual Studio Code.
+
+However, to debug the firmware running on the network core, you need to set up separate debug sessions for each core:
+
+* Application core.
+* Network core.
+
+Perform the following steps to start a debugging session in network core:
```
Comment (peknis) [inline doc/nrf/ug_nrf5340.rst:637]:
```suggestion
Complete the following steps to start a debugging session in network core:
```

**Review by peknis — CHANGES_REQUESTED** [review]

**Review by peknis — APPROVED** [review]

**Comment by grochu** [comment]
Hi @MarekPieta, could you also take a look at this description or re-assign to someone in your team? I think you have spent significant time with nRF5340.

**Comment by MarekPieta** [comment]
Hi @grochu
I will try to take a look at this one

**Review by MarekPieta — APPROVED** [review]
Looks good. Please keep in mind that I do not have practical experience with debugging in SES nor VS Code (I use Ubuntu and CLI in most cases).

**Review by crfosse — APPROVED** [review]
Looks good to me! The way GPIO is assigned will probably change with the [PR #42576](https://github.com/zephyrproject-rtos/zephyr/pull/42576) in Zephyr, but guess the approach described here will remain viable.

**Review by divipillai — APPROVED** [review]

**Review by trond-snekvik — APPROVED** [review]

**Pull request merged**